pkg/util/cloudfoundry/bbscache.go

```diff
@@ -192,39 +212,83 @@ func (bc *BBSCache) readData() {
 	bc.Lock()
 	defer bc.Unlock()
 	log.Debug("Data from BBS API read successfully, refreshing the cache")
-	bc.actualLRPs = actualLRPs
+	bc.actualLRPsByAppGUID = actualLRPsByAppGUID
+	bc.actualLRPsByCellID = actualLRPsByCellID
 	bc.desiredLRPs = desiredLRPs
 	bc.lastUpdated = time.Now()
 	bc.pollSuccesses++
 }
 
-func (bc *BBSCache) readActualLRPs() (map[string][]ActualLRP, error) {
-	actualLRPs := map[string][]ActualLRP{}
+func (bc *BBSCache) readActualLRPs() (map[string][]*ActualLRP, map[string][]*ActualLRP, error) {
+	actualLRPsByAppGUID := map[string][]*ActualLRP{}
+	actualLRPsByCellID := map[string][]*ActualLRP{}
 	actualLRPsBBS, err := bc.bbsAPIClient.ActualLRPs(bc.bbsAPIClientLogger, models.ActualLRPFilter{})
 	if err != nil {
-		return actualLRPs, err
+		return actualLRPsByAppGUID, actualLRPsByCellID, err
 	}
 	for _, lrp := range actualLRPsBBS {
 		alrp := ActualLRPFromBBSModel(lrp)
-		if lrpList, ok := actualLRPs[alrp.AppGUID]; ok {
-			actualLRPs[alrp.AppGUID] = append(lrpList, alrp)
+		if _, ok := actualLRPsByAppGUID[alrp.AppGUID]; ok {
+			actualLRPsByAppGUID[alrp.AppGUID] = append(actualLRPsByAppGUID[alrp.AppGUID], &alrp)
+			if _, ok := actualLRPsByCellID[alrp.CellID]; ok {
+				actualLRPsByCellID[alrp.CellID] = append(actualLRPsByCellID[alrp.CellID], &alrp)
+			} else {
+				actualLRPsByCellID[alrp.CellID] = []*ActualLRP{&alrp}
+			}
 		} else {
-			actualLRPs[alrp.AppGUID] = []ActualLRP{alrp}
+			actualLRPsByAppGUID[alrp.AppGUID] = []*ActualLRP{&alrp}
+			if _, ok := actualLRPsByCellID[alrp.CellID]; ok {
+				actualLRPsByCellID[alrp.CellID] = append(actualLRPsByCellID[alrp.CellID], &alrp)
+			} else {
+				actualLRPsByCellID[alrp.CellID] = []*ActualLRP{&alrp}
+			}
 		}
 	}
 	log.Debugf("Successfully read %d Actual LRPs", len(actualLRPsBBS))
-	return actualLRPs, nil
+	return actualLRPsByAppGUID, actualLRPsByCellID, nil
 }
 
-func (bc *BBSCache) readDesiredLRPs() ([]DesiredLRP, error) {
+func (bc *BBSCache) readDesiredLRPs() (map[string]*DesiredLRP, error) {
 	desiredLRPsBBS, err := bc.bbsAPIClient.DesiredLRPs(bc.bbsAPIClientLogger, models.DesiredLRPFilter{})
 	if err != nil {
-		return []DesiredLRP{}, err
+		return map[string]*DesiredLRP{}, err
 	}
-	desiredLRPs := make([]DesiredLRP, len(desiredLRPsBBS))
-	for i, lrp := range desiredLRPsBBS {
-		desiredLRPs[i] = DesiredLRPFromBBSModel(lrp)
+	desiredLRPs := make(map[string]*DesiredLRP, len(desiredLRPsBBS))
+	for _, lrp := range desiredLRPsBBS {
+		desiredLRP := DesiredLRPFromBBSModel(lrp)
+		desiredLRPs[desiredLRP.AppGUID] = &desiredLRP
 	}
 	log.Debugf("Successfully read %d Desired LRPs", len(desiredLRPsBBS))
 	return desiredLRPs, nil
 }
+
+// ExtractTags extract all the container tags for each app in the cache running on the specified node
+// and returns a mapping of tags by instance GUID
+func (bc *BBSCache) ExtractTags(nodename string) map[string][]string {
```

Comment:
Do you think it would make more sense to have this pre-computed for all cells in the same way the actual/desired LRPs are and then just return result? Doing all this dynamically with potentially thousands of nodes polling in short interval could be very CPU-intensive when recomputing all the time.